### DISCUSSION TITLE
Getting the app.xml file right. Assets.car. Launchscreen.storyboardc and more!

### DISCUSSION THREAD

**Discussion by little-fat-girl** [airsdk/Adobe-Runtime-Support#2896] [General]
So, it's been a while since I deployed anything to the AppStore (about 9 years). And I remember my app.xml file being pretty complicated and having to make it read-only so stuff didn't get overwritten when I compiled out of Animate. The autogenerated app.xml file I'm seeing these days looks waaaaaay simpler.  It's this:

```
<?xml version="1.0" encoding="UTF-8" standalone="no" ?>
<application xmlns="http://ns.adobe.com/air/application/50.2">
  <id>com.mycompany.myapp</id>
  <versionNumber>1.4</versionNumber>
  <filename>App Name</filename>
  <description/>
```
  
Things seem to deploy just fine to my local devices (which is awesome). But I'm getting ready to deploy to the AppStore and want to make sure I have this thing right so I don't upset the good people at Apple. 

I did the asset.car and Launchscreen.storyboardc on my ancient mac mini. That's new to me. Last time I had a bunch of icons that were listed in the app.xml file. I still have those files listed in my publish settings in Adobe Animate. I had seen somewhere that this helps compatibility with old versions and I would love for this to be as backward compatible as possible.

So my questions are these:

1. How do I know that my launchscreen.storyboardc file and assets.car are being used in the new build? Do I need to list those files in my publish settings in Animate (like the default splashscreen images or the icon files)? Or do I need to manually add those items to the app.xml?

I had seen a tutorial by distriqt where they added it to the app.xml thusly:

> Once you have completed your storyboard and added the application icons you can add these to your AIR applicaiton.
> 
> Copy Assets.car and LaunchScreen.storyboardc and package with AIR application at the root level of your application.
> 
> Add the following to your info additions in your application descriptor:
> 

> ```
> <key>UILaunchStoryboardName</key>
> <string>LaunchScreen</string>
> 
> <iPhone>
>     <InfoAdditions><![CDATA[            
>         
>         <key>UILaunchStoryboardName</key>
>         <string>LaunchScreen</string>
>     ]]></InfoAdditions>
> </iPhone>!-- ... -->
>         <key>UILaunchStoryboardName</key>
>         <string>LaunchScreen</string>
>     ]]></InfoAdditions>
> </iPhone>
> ```

2. Are there more things I need to set here? 

Do I need to set the Minimum iOS version?  I would love for it to be whatever the lowest number is to maximize compatibility.  Does the current Air build default to something higher?  

Do I need to set anything to make sure I am deploying for iPhone and iPad? I have this set in the publishing settings in Animate, but I want to make sure.

I also had stuff in the old app.xml that looked like this:

```
  <initialWindow>
    <content>km-cc.swf</content>
    <systemChrome>standard</systemChrome>
    <transparent>false</transparent>
    <visible>true</visible>
    <fullScreen>true</fullScreen>
    <aspectRatio>landscape</aspectRatio>
    <renderMode>gpu</renderMode>
    <autoOrients>true</autoOrients>
  </initialWindow>
  <icon>
    <image29x29>AppIcon29x29.png</image29x29>
    <image58x58>AppIcon29x29@2x.png</image58x58>
    <image87x87>AppIcon29x29@3x.png</image87x87>
    <image512x512>iTunesArtwork.png</image512x512>
    <image1024x1024>iTunesArtwork@2x.png</image1024x1024>
    <image57x57>AppIcon57x57.png</image57x57>
    <image114x114>AppIcon57x57@2x.png</image114x114>
    <image72x72>AppIcon72x72.png</image72x72>
    <image50x50>AppIcon50x50.png</image50x50>
    <image100x100>AppIcon50x50@2x.png</image100x100>
    <image144x144>AppIcon72x72@2x.png</image144x144>
    <image60x60>AppIcon60x60.png</image60x60>
    <image120x120>AppIcon60x60@2x.png</image120x120>
    <image180x180>AppIcon60x60@3x.png</image180x180>
    <image76x76>AppIcon76x76.png</image76x76>
    <image152x152>AppIcon76x76@2x.png</image152x152>
  </icon>
  <customUpdateUI>false</customUpdateUI>
  <allowBrowserInvocation>false</allowBrowserInvocation>
  <iPhone>
    <InfoAdditions>
      <![CDATA[<key>UIDeviceFamily</key>
        <array>
        <string>1</string>
        <string>2</string>
        </array>
```
Do I need any of that?  And, if this is listed in a tutorial somewhere, please point to that and I'll ask smarter questions.

Thanks for the help!

**Comment by ajwfrost** [comment]
Hi

Your initial XML file seems very brief, I would expect the `<initialWindow>` section to be needed at least! But coming to that..

Q1 -> yes you just need to package in the Assets.car and LaunchScreen.storyboard at the root of your application, and add the storyboard filename root to the app descriptor InfoAdditions per that tutorial from Distriqt.

Worth checking the output of the ADT packaging tool to see if it gives any warnings when it packages. For example if you haven't got a commercial license for AIR SDK, then we will ignore your storyboard and put our own one in instead - at which point, the Assets.car file needs to contain the right images otherwise the launch screen is a couple of seconds of blankness...

Q2 -> a few questions! Minimum OS is set to 11 at the moment I think, by default, but could be overridden by requests from ANEs. But generally we just set it to whatever we have to based on Apple/Xcode requirements. To ensure it's shown for both iPhones and iPads, use that `UIDeviceFamily` key per your sample there, and include both families (1 and 2). For app store, you'll need the icons that you mention there, I had a feeling it didn't want the 1024-pixel one as that was meant to be provided separately? But I think/hope the package would still work and would exclude it from the generated ICNS file .. again, worth checking for the details of any warnings when it builds.

Seems to be pretty much there though I think. Code-signing is something you'll have to go through of course; the best option to ensure that this works is to use JDK 11 or similar, I think it's from JDK 17 onwards that it all breaks..

thanks

**Reply by little-fat-girl** [reply]
Thanks for the quick reply. I have time to work on this today, so this really helps. 

I've got this thing working now and deploying to my iPhone and iPad straight out of Animate. So that's good news. 

Good to know that I need to list all this stuff out in the app.xml. Are there any templates out there that list the various variables and their meanings? I would assume the the on I used works for my app, but it is 9 years old and I wonder if there aren't more items needed.  Back then, Animate auto-generated that thing.  And it got it mostly right. It was definitely more complete than what I'm getting today.

Also, you mentioned Code-signing and JDK 11.  This stuff is handled for me on ios by Animate and your SDK, right? If I'm not doing android, I don't need that, do I?

**Reply by little-fat-girl** [reply]
Ha!  Ok. So when I went to edit my app.xml file, I found a whole bunch more stuff further down in the document.  There were just a bunch of carriage returns that made the page look blank.  So, now I'm sorting through that to see if it answers some questions....

**Reply by little-fat-girl** [reply]
Also, found a breakdown of the "application descriptor file": https://help.adobe.com/en_US/air/build/WS5b3ccc516d4fbf351e63e3d118666ade46-7f84.html

**Reply by ajwfrost** [reply]
We're updating that documentation and hosting it at https://airsdk.dev/docs/building/application-descriptor-files/introduction

Templates and rules for the files are in the AIR SDK under the "templates\air" folder, "descriptor-template.xml" which is used by some IDE tools to initially create an app descriptor file, and "Descriptor.50.2.xsd" as the formal schema definition.

Code-signing should be handled by Animate if you've put your certificate information into that. But ultimately that pushes the request in to ADT which runs in Java and so if that Java version is too recent, we lose access to the classes that do some of the security algorithms needed for signing. But with Animate I believe it uses its own deployment of the Java runtime so you should be okay.

thanks
